### ISSUE TITLE
Trial best score when pruned

### ISSUE THREAD

**Issue by geoffsmith** [optuna/optuna#2575]
Hi!

I'm using optuna to tune a neural net where for some hyper-parameters, the model overfits quite quickly (e.g. larger capacity /  higher learning rate). I've added early stopping (using pytorch-lightning) to stop early if the val_loss is getting worse after several epochs. Because of this, I'm returning the **best** validation loss from the objective function, not the last one. However, I've noticed that if I use the optuna pruning as well, then the last loss will be used rather than the best loss.

Overall, it's not necessarily a big deal because I can look at the plots and see which trial was best. However, I'm wondering if the score is affecting the TPE sampler b/c it thinks that the pruned trials performed way worse than they actually did?

And if so, what are my options? Perhaps there's a way to configure optuna to consider the best score, not just the last one. Or I could perhaps handle the `TrialPruned` exception in my objective function and return the best score (but I'm not sure if that will cause problems with the pruner itself?) Or maybe the sample is already using the best score?

Any tips much appreciated :)

**Comment by fsmosca** [comment]
> However, I've noticed that if I use the optuna pruning as well, then the last loss will be used rather than the best loss.

Which pruner did you use?

**Comment by geoffsmith** [comment]
HyperBand

**Comment by fsmosca** [comment]
[Revised example](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.pruners.HyperbandPruner.html#optuna.pruners.HyperbandPruner) from the doc.

```python
import numpy as np
from sklearn.datasets import load_iris
from sklearn.linear_model import SGDClassifier
from sklearn.model_selection import train_test_split

import optuna

X, y = load_iris(return_X_y=True)
X_train, X_valid, y_train, y_valid = train_test_split(X, y)
classes = np.unique(y)
n_train_iter = 100

num_pruned = 0

def objective(trial):
    global num_pruned
    alpha = trial.suggest_float("alpha", 0.0, 1.0)
    clf = SGDClassifier(alpha=alpha)

    for step in range(n_train_iter):
        clf.partial_fit(X_train, y_train, classes=classes)

        intermediate_value = clf.score(X_valid, y_valid)
        trial.report(intermediate_value, step)

        if trial.should_prune():
            num_pruned += 1
            print(f'trial: {trial.number}, step: {step}, pruned intermediate value: {intermediate_value}, alpha: {alpha}')
            print(f'total trials pruned: {num_pruned}, trials not pruned: {1 + trial.number - num_pruned}')
            raise optuna.TrialPruned()

    return clf.score(X_valid, y_valid)


study = optuna.create_study(
    direction="maximize",
    pruner=optuna.pruners.HyperbandPruner(
        min_resource=1, max_resource=n_train_iter, reduction_factor=3
    ),
)
study.optimize(objective, n_trials=30)

print(f'best trial: {study.best_trial.number}, best param: {study.best_params}, best_value: {study.best_value}')
```

### Sample output:
Here the best trial is 6. This is the trial given when we query the best trial after the study of 30 trials.

`best trial: 6, best param: {'alpha': 0.14818617338562634}, best_value: 0.7105263157894737`

From the log this looks normal to me.

Could you explain more about this?
`then the last loss will be used rather than the best loss`

```c++
[I 2021-04-11 00:44:54,784] A new study created in memory with name: no-name-6310fbbf-608a-4da8-800b-4625de91e5e7
[I 2021-04-11 00:44:55,179] Trial 0 finished with value: 0.631578947368421 and parameters: {'alpha': 0.6118039205201573}. Best is trial 0 with value: 0.631578947368421.
[I 2021-04-11 00:44:55,565] Trial 1 finished with value: 0.631578947368421 and parameters: {'alpha': 0.1844286186641474}. Best is trial 0 with value: 0.631578947368421.
trial: 2, step: 3, pruned intermediate value: 0.631578947368421, alpha: 0.14164838457584306
total trials pruned: 1, trials not pruned: 2
[I 2021-04-11 00:44:55,581] Trial 2 pruned. 
[I 2021-04-11 00:44:55,964] Trial 3 finished with value: 0.631578947368421 and parameters: {'alpha': 0.3677193339212349}. Best is trial 0 with value: 0.631578947368421.
[I 2021-04-11 00:44:55,979] Trial 4 pruned. 
trial: 4, step: 3, pruned intermediate value: 0.631578947368421, alpha: 0.48093746324367703
total trials pruned: 2, trials not pruned: 3
trial: 5, step: 3, pruned intermediate value: 0.631578947368421, alpha: 0.869783794826699
total trials pruned: 3, trials not pruned: 3
[I 2021-04-11 00:44:55,996] Trial 5 pruned. 
[I 2021-04-11 00:44:56,384] Trial 6 finished with value: 0.7105263157894737 and parameters: {'alpha': 0.14818617338562634}. Best is trial 6 with value: 0.7105263157894737.
[I 2021-04-11 00:44:56,763] Trial 7 finished with value: 0.631578947368421 and parameters: {'alpha': 0.5608533697408314}. Best is trial 6 with value: 0.7105263157894737.
[I 2021-04-11 00:44:57,178] Trial 8 finished with value: 0.631578947368421 and parameters: {'alpha': 0.9738162241091641}. Best is trial 6 with value: 0.7105263157894737.
trial: 9, step: 3, pruned intermediate value: 0.631578947368421, alpha: 0.633683828392538
total trials pruned: 4, trials not pruned: 6
[I 2021-04-11 00:44:57,196] Trial 9 pruned. 
trial: 10, step: 81, pruned intermediate value: 0.631578947368421, alpha: 0.940893216577942
total trials pruned: 5, trials not pruned: 6
[I 2021-04-11 00:44:57,502] Trial 10 pruned. 
[I 2021-04-11 00:44:57,873] Trial 11 finished with value: 0.631578947368421 and parameters: {'alpha': 0.8509139367885137}. Best is trial 6 with value: 0.7105263157894737.
trial: 12, step: 81, pruned intermediate value: 0.631578947368421, alpha: 0.5773174168675299
total trials pruned: 6, trials not pruned: 7
[I 2021-04-11 00:44:58,196] Trial 12 pruned. 
[I 2021-04-11 00:44:58,566] Trial 13 finished with value: 0.631578947368421 and parameters: {'alpha': 0.4627796885502774}. Best is trial 6 with value: 0.7105263157894737.
[I 2021-04-11 00:44:58,925] Trial 14 finished with value: 0.631578947368421 and parameters: {'alpha': 0.19621286577524788}. Best is trial 6 with value: 0.7105263157894737.
[I 2021-04-11 00:44:59,329] Trial 15 finished with value: 0.631578947368421 and parameters: {'alpha': 0.4306869604715027}. Best is trial 6 with value: 0.7105263157894737.
trial: 16, step: 81, pruned intermediate value: 0.631578947368421, alpha: 0.7852474754404934
total trials pruned: 7, trials not pruned: 10
[I 2021-04-11 00:44:59,702] Trial 16 pruned. 
[I 2021-04-11 00:45:00,079] Trial 17 finished with value: 0.631578947368421 and parameters: {'alpha': 0.30465712739672146}. Best is trial 6 with value: 0.7105263157894737.
[I 2021-04-11 00:45:00,547] Trial 18 finished with value: 0.7105263157894737 and parameters: {'alpha': 0.07914565504657389}. Best is trial 6 with value: 0.7105263157894737.
trial: 19, step: 1, pruned intermediate value: 0.631578947368421, alpha: 0.5126313701548884
total trials pruned: 8, trials not pruned: 12
[I 2021-04-11 00:45:00,557] Trial 19 pruned. 
[I 2021-04-11 00:45:01,011] Trial 20 finished with value: 0.631578947368421 and parameters: {'alpha': 0.595100433702487}. Best is trial 6 with value: 0.7105263157894737.
[I 2021-04-11 00:45:01,390] Trial 21 finished with value: 0.631578947368421 and parameters: {'alpha': 0.6422803532419401}. Best is trial 6 with value: 0.7105263157894737.
trial: 22, step: 1, pruned intermediate value: 0.631578947368421, alpha: 0.0638582278947804
total trials pruned: 9, trials not pruned: 14
[I 2021-04-11 00:45:01,406] Trial 22 pruned. 
[I 2021-04-11 00:45:01,799] Trial 23 finished with value: 0.631578947368421 and parameters: {'alpha': 0.6958854596080957}. Best is trial 6 with value: 0.7105263157894737.
[I 2021-04-11 00:45:02,202] Trial 24 finished with value: 0.631578947368421 and parameters: {'alpha': 0.703220854845348}. Best is trial 6 with value: 0.7105263157894737.
trial: 25, step: 9, pruned intermediate value: 0.631578947368421, alpha: 0.9838028186373349
total trials pruned: 10, trials not pruned: 16
[I 2021-04-11 00:45:02,244] Trial 25 pruned. 
[I 2021-04-11 00:45:02,677] Trial 26 finished with value: 0.631578947368421 and parameters: {'alpha': 0.6629960940841575}. Best is trial 6 with value: 0.7105263157894737.
trial: 27, step: 1, pruned intermediate value: 0.39473684210526316, alpha: 0.00921974479095497
total trials pruned: 11, trials not pruned: 17
[I 2021-04-11 00:45:02,695] Trial 27 pruned. 
[I 2021-04-11 00:45:03,050] Trial 28 finished with value: 0.631578947368421 and parameters: {'alpha': 0.6167080676763365}. Best is trial 6 with value: 0.7105263157894737.
trial: 29, step: 1, pruned intermediate value: 0.631578947368421, alpha: 0.622025865395947
total trials pruned: 12, trials not pruned: 18
best trial: 6, best param: {'alpha': 0.14818617338562634}, best_value: 0.7105263157894737
[I 2021-04-11 00:45:03,066] Trial 29 pruned. 

```

**Comment by geoffsmith** [comment]
Thanks for taking the time! Yes, so say you have these intermediate values for a particular trial:

Step 1: 0.8
Step 2: 0.7
Step 3: 0.6 <- Best 
Step 4: 0.65
Step 5: 0.7 <- Pruned

If this gets pruned at step 5, my guess is that the sampler will consider the trial to have a value of 0.7 and use this loss in selecting future samples. What I would prefer is for the trial to report a value of 0.6 (the best one, from step 3) instead. E.g. this part of the hyper-parameter space is probably more promising that it would seem if you took the value from the final step.

Hope that make more sense :)

**Comment by fsmosca** [comment]
> If this gets pruned at step 5, my guess is that the sampler will consider the trial to have a value of 0.7

No, it is pruned at this particular trial with corresponding alpha value.

The optimizer will only start to prune when it already have some trials that were done. Once it has some alpha and loss values, it will then consider pruning trials. Those intermediate values that are not good compared to the loss of past trials would cause the trial to be pruned.

Let's have an example, direction=minimize, say there are 100 samples to step into and 30 trials to find best param alpha.
```
trial 0, alpha 0.25, alpha is suggested by the optimizer
    step 1 loss 0.125
    step 2 loss 0.126
    ...
    step 100 loss 0.124  -> our objective value for this trial, send this to optimizer

trial 1, alpha 0.26
    step 1 loss 0.119
    ...
    step 100 loss 0.120
   
trial 2, alpha 0.30
    step 1 loss 0.5
        Prune trial 2 because the intermediate loss at step 1 (0.5) is way
        higher than the loss in trial 1 (0.120). The alpha value of 0.30 is bad.
        The sampler will not consider 0.5, but instead the sampler will consider alpha 0.30
        as bad and its probability to be selected in future trials is very low or zero.

If we query at this particular stage which loss is best, it will send us 0.120 from trial 1 with alpha 0.26.

trial 3, alpha ...

...

trial 29, alpha ...
    step 1 loss ...
    ...
    step 100 loss ...
```

**Comment by fsmosca** [comment]
> What I would prefer is for the trial to report a value of 0.6 (the best one, from step 3) instead.

Could be unsafe as you have not yet completed evaluating the samples? You just completed 3 samples out of 100. However if you really want to return early just break from the step loop and return 0.6.

**Comment by geoffsmith** [comment]
Sorry, I don't think I'm explaining myself very well... say for a given trial, the loss follows a U-shape, then it may get pruned at some point, even though a previous intermediate loss was the best of all trials. E.g.

Best loss of trials 1 and 2 is 0.120, then on trial 3 with alpha=0.3
Step 1: 0.125
Step 2: 0.12
Step 3: 0.11 <- Best so far, this is the sweet spot before the neural net starts overfitting
Step 4: 0.12
Step 5: 0.5 <- V bad, gets pruned, my neural net is now overfitting heavily

In this case I would like the sampler to consider alpha=0.3 very good instead of very bad.

**Comment by fsmosca** [comment]
The sampler is only after the optimization of alpha, you can return 0.11 for alpha 0.3 if we know it is U-shape. Or don't use pruner and finish all the steps then pick the lowest loss and return it.

**Comment by Crissman** [comment]
Thanks @fsmosca . Looks like this question is answered, so let me close it.

Feel free to open if not answered or ask on Gitter.

**Comment by l-cdc** [comment]
I am having the same issue, and I do not have the impression that it has been answered. I find @fsmosca's last comment ambiguous – depending on the interpretation, it might or might not agree with the following.

Overall, the key issue is that there is a mismatch between what is reported to optuna (usually the epoch validation loss), and the "actual" performance, which is given by the best performance across all epochs in the current trial, usually not associated to the last epoch. This is, as @geoffsmith points out, typical for DL models, because they tend to overfit at some point.

Hence, sampling will be impacted, and – when enabled – the pruning logic might curtail some actually promising trials.

The most obvious fix is to report the best loss instead of the epoch loss to optuna, both at the end of each epoch (as _intermediate_ values) and as the result of the optimisation function. Depending on the framework, this might mean adjusting the integration code e.g., for [PyTorch Lightning](https://optuna.readthedocs.io/en/v1.2.0/_modules/optuna/integration/pytorch_lightning.html) to offer this option, if not making it the default.

Another option is to return the best loss from the optimisation function only (as @fsmosca seems to indicate), but the caveat is that the intermediate values will be inconsistent with the final value. This might work better for pruning, though I know currently too little to comment. However, since some samplers [consider also pruned trials](https://github.com/optuna/optuna/issues/1647), and the [value for pruned trials is set to the last intermediate value](https://github.com/optuna/optuna/blob/83c8022bcb3847b261e3f0b65aeaccf46dbb8cb6/optuna/study/_tell.py#L163) this might yield less than ideal results. To address this, samplers should be set to ignore pruned trials if needed, or pruning should be disabled (and maybe other early stopping mechanisms could be used instead).

Any comments? I will try out the former approach and submit a PR if it works.

**Comment by evfro** [comment]
I also feel that the original question was not unanswered.
@geoffsmith what do you think about the following workaround: during the intermediate steps, `report the current best value` instead of the latest one.

Let's say, your series of intermediate values are [0.3, 0.5, 0.2, 0.3, 0.4]. You would report [0.3, 0.3, 0.2, 0.2, 0.2]. That way, the pruner is still considering your trial and will prune it if it's really bad. And if not, the trial won't be pruned and it will be used by the sampler for making decision on the next step in the search space.

The main drawback of this workaround is that you'll lose the actual values of intermediate steps coming after the best trial, so you won't be able to use standard plots for analysis of the learning curves as they will only show the `running best` not the actual values. You could still save the real scores in the `user_attributes`, but then it would require manually extracting this info and making the necessary analysis with your own code.
